apps/angular/crud/src/app/components/to-do/to-do.component.spec.ts

```diff
@@ -0,0 +1,143 @@
+    expect(component).toBeTruthy();
+  });
+
+  it('should initialize component and call setTodo',  () => {
+     component.ngOnInit();
+    expect(toDoStoreMock.setTodo).toHaveBeenCalled();
+  });
+
+  it('should render todo items correctly', async () => {
+    const mockToDoList = [
+      {
+        completed: false,
+        id: 101,
+        title: 'explicabo enim cumque porro aperiam occaecati minima',
+        userId: 6,
+      },
+      {
+        completed: false,
+        id: 102,
+        title: 'explicabo enim cumque porro aperiam occaecati minima',
+        userId: 6,
+      },
+    ];
+    toDoStoreMock.todo.mockReturnValue(mockToDoList);
+    fixture.detectChanges();
+
+    const listItems = fixture.nativeElement.querySelectorAll('.list-items');
+    expect(listItems.length).toBe(2);
+  });
+
+  it('should update todo item when update button is clicked', async () => {
+    const mockItem = {
+      completed: false,
+      id: 102,
+      title: 'explicabo enim cumque porro aperiam occaecati minima',
+      userId: 6,
+    };
+    toDoStoreMock.todo.mockReturnValue([mockItem]);
+    fixture.detectChanges();
+
+    const updateButton = fixture.nativeElement.querySelector('.update');
+    updateButton.click();
+
+    expect(toDoStoreMock.updateTodo).toHaveBeenCalledWith({
+      ...mockItem,
+      title: expect.any(String),
+    });
+  });
+
+  it('should delete todo item when delete button is clicked', async () => {
+    const mockItem = {
+      completed: false,
+      id: 102,
+      title: 'explicabo enim cumque porro aperiam occaecati minima',
+      userId: 6,
+    };
+    toDoStoreMock.todo.mockReturnValue([mockItem]);
+    const deleteButton = fixture.nativeElement.querySelector('.delete');
+    deleteButton.click();
+
+    expect(toDoStoreMock.updateTodo).toHaveBeenCalledWith(mockItem.id);
+  });
+
+  it('should disable button when spinner is loading', async () => {
+    toDoStoreMock.isLoading.mockReturnValue(true);
+    fixture.detectChanges();
+
+    const updateButton =
+      fixture.nativeElement.querySelector('button[mat-button]');
+    const deleteButton = fixture.nativeElement.querySelector(
+      'button[color="warn"]'
+    );
+    const spinner = fixture.nativeElement.querySelector('mat-spinner');
+
+    expect(updateButton.disabled).toBe(true);
+    expect(deleteButton.disabled).toBe(true);
+    expect(spinner).toBeTruthy();
+  });
+
+  it('should enable button and hide spinner when spinner is not loading', async () => {
+    toDoStoreMock.isLoading.mockReturnValue(false);
+    fixture.detectChanges();
+
+    const updateButton =
+      fixture.nativeElement.querySelector('button[mat-button]');
+    const deleteButton = fixture.nativeElement.querySelector(
+      'button[color="warn"]'
+    );
+    const spinner = fixture.nativeElement.querySelector('mat-spinner');
+
+    expect(updateButton.disabled).toBe(false);
+    expect(deleteButton.disabled).toBe(false);
+    expect(spinner).toBeFalsy();
+  });
+});
```

Comment:
@tomalaforge  "Without Implementation details" , this is some great information, never seen it that way,
I like to have a conversation around testing, enlighten my self and clear my doubts

In testing library,
sometimes we have to select based on their position (i.e), i am constantly thinking what if the position changes, the tests will break
if i can select with their attributes that doesnt change, it would be great, 
is that possible with testing library ? i am still on the midst of testing challenges

Reply:
With testing lib, you use a11y selector to get your component, not position or any random attribut. So you are testing a11y at the same time and you are not relying on angular internal API. 
Try the testing challenges, you should get a better understanding of testing lib and cypress

Reply:
Yes, i have done some challenges and i get the idea of not relying on angular internal api for testing, 
thanks for the information!
I had fun on learning while doing testing challenges as well